File: tests/solutions_js/pacific_atlantic_water_flow.js
class Solution {
    pacificAtlantic(heights) {
        if (!heights || heights.length === 0 || heights[0].length === 0) return [];

        const rows = heights.length;
        const cols = heights[0].length;
        const directions = [[0, 1], [0, -1], [1, 0], [-1, 0]];

        const bfs = (starts) => {
            const visited = new Set(starts.map(([r, c]) => `${r},${c}`));
            const queue = [...starts];
            let head = 0;
            while (head < queue.length) {
                const [r, c] = queue[head++];
                for (const [dr, dc] of directions) {
                    const nr = r + dr;
                    const nc = c + dc;
                    const key = `${nr},${nc}`;
                    if (
                        nr >= 0 && nr < rows &&
                        nc >= 0 && nc < cols &&
                        !visited.has(key) &&
                        heights[nr][nc] >= heights[r][c]
                    ) {
                        visited.add(key);
                        queue.push([nr, nc]);
                    }
                }
            }
            return visited;
        };

        const pacificStarts = [];
        const atlanticStarts = [];
        for (let r = 0; r < rows; r++) {
            pacificStarts.push([r, 0]);
            atlanticStarts.push([r, cols - 1]);
        }
        for (let c = 0; c < cols; c++) {
            pacificStarts.push([0, c]);
            atlanticStarts.push([rows - 1, c]);
        }

        const pacific = bfs(pacificStarts);
        const atlantic = bfs(atlanticStarts);

        const result = [];
        for (const key of pacific) {
            if (atlantic.has(key)) {
                result.push(key.split(",").map(Number));
            }
        }

        result.sort((a, b) => a[0] - b[0] || a[1] - b[1]);
        return result;
    }
}

File: tests/solutions_cpp/pacific_atlantic_water_flow.cpp
#include "json_helper.h"

class Solution {
public:
    vector<vector<int>> pacificAtlantic(vector<vector<int>>& heights) {
        vector<vector<int>> result;
        if (heights.empty() || heights[0].empty()) return result;
        int rows = heights.size(), cols = heights[0].size();
        vector<vector<bool>> pac(rows, vector<bool>(cols, false));
        vector<vector<bool>> atl(rows, vector<bool>(cols, false));
        for (int r = 0; r < rows; r++) {
            dfs(heights, r, 0, pac, rows, cols);
            dfs(heights, r, cols - 1, atl, rows, cols);
        }
        for (int c = 0; c < cols; c++) {
            dfs(heights, 0, c, pac, rows, cols);
            dfs(heights, rows - 1, c, atl, rows, cols);
        }
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (pac[r][c] && atl[r][c]) result.push_back({r, c});
            }
        }
        sort(result.begin(), result.end());
        return result;
    }
private:
    void dfs(vector<vector<int>>& h, int r, int c, vector<vector<bool>>& visited, int rows, int cols) {
        if (visited[r][c]) return;
        visited[r][c] = true;
        int dr[] = {0, 0, 1, -1};
        int dc[] = {1, -1, 0, 0};
        for (int i = 0; i < 4; i++) {
            int nr = r + dr[i], nc = c + dc[i];
            if (nr >= 0 && nr < rows && nc >= 0 && nc < cols &&
                !visited[nr][nc] && h[nr][nc] >= h[r][c]) {
                dfs(h, nr, nc, visited, rows, cols);
            }
        }
    }
};

int main() {
    auto args = parse_args();
    Solution sol;
    auto heights = to_vector_vector_int(args[0]);
    auto result = sol.pacificAtlantic(heights);
    cout << to_json(result) << endl;
    return 0;
}
